src/main/java/org/truffleruby/core/array/ArrayBuilderNode.java

```diff
@@ -25,519 +30,222 @@
+            return AppendOneNodeGen.create(context);
         }
 
-        @Override
-        public Object start() {
-            return new double[expectedLength];
-        }
+        private final RubyContext context;
 
-        @Override
-        public Object start(int length) {
-            if (length > expectedLength) {
-                CompilerDirectives.transferToInterpreterAndInvalidate();
-                return restart(length, length + " > " + expectedLength + " (expected)");
-            }
-
-            return new double[expectedLength];
-        }
-
-        @Override
-        public Object ensure(Object store, int length) {
-            if (length > ((double[]) store).length) {
-                CompilerDirectives.transferToInterpreterAndInvalidate();
-                return ensureFallback(store, length);
-            }
-            return store;
+        public AppendOneNode(RubyContext context) {
+            this.context = context;
         }
 
-        @Override
-        public Object appendArray(Object store, int index, DynamicObject array) {
-            Object otherStore = Layouts.ARRAY.getStore(array);
-
-            if (otherStore == null) {
-                return store;
-            }
-
-            if (otherDoubleStoreProfile.profile(otherStore instanceof double[])) {
-                System.arraycopy(otherStore, 0, store, index, Layouts.ARRAY.getSize(array));
-                return store;
-            }
-
-            CompilerDirectives.transferToInterpreterAndInvalidate();
-            return replace(new ObjectArrayBuilderNode(expectedLength)).
-                    appendArray(ArrayUtils.box((double[]) store), index, array);
-        }
+        public abstract Object executeAppend(Object array, int index, Object value);
 
-        @Override
-        public Object appendValue(Object store, int index, Object value) {
-            if (store instanceof double[] && value instanceof Double) {
-                ((double[]) store)[index] = (double) value;
-                return store;
-            } else {
-                CompilerDirectives.transferToInterpreterAndInvalidate();
-                return appendValueFallback(store, index, value, expectedLength);
+        @Specialization(guards = { "strategy.matchesStore(array)", "strategy.accepts(value)" },
+                        limit = "1")
+        public Object appendCompatibleType(Object array, int index, Object value,
+                @Cached("ofStore(array)") ArrayStrategy strategy,
+                @Cached("createBinaryProfile()") ConditionProfile expandProfile) {
+            ArrayMirror mirror = strategy.newMirrorFromStore(array);
+            if (expandProfile.profile(index >= mirror.getLength())) {
+                final int capacity = ArrayUtils.capacityForOneMore(context, mirror.getLength());
+                replaceNodes(strategy, capacity);
+                mirror = mirror.copyArrayAndMirror(capacity);
```

Comment:
While it's nice in the code to do this, PEA doesn't handle it well iirc.
So in this case, it's much better to manually duplicate the tail and never re-assign `mirror` (same below).

Reply:
Ah, maybe it doesn't matter in this case since `replaceNodes` is a transfer.
I think there should be an explicit transfer at the start of the branch, and then we also don't need the ConditionProfile.

Reply:
Okay, I'll move the transfers to the users of `replaceNodes`.